File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # mp = {}

        # l = 0

        # res = 0

        # for r in range(len(s)):
        #     if s[r] in mp:
        #         l = max(mp[s[r]] + 1, l)
            
        #     mp[s[r]] = r
        #     res = max(res, r - l + 1)

        # return res

        l = 0
        r = 0
        length = 0
        tracker = set()
        while r < len(s):
            while s[r] in tracker:
                tracker.remove(s[l])
                l+=1

            tracker.add(s[r])

            length = max(length, r - l+1)
            r+=1

        return length

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        
        // we need to give the longest window as output
        // brute force approach is O(n^2)

        unordered_map<int,int> count;
        int res = 0;

        int l = 0 ; 
        int maxFrequency = 0;

        for (int r = 0 ; r < s.size() ; r++){
            count[s[r]]++;
            maxFrequency = max(maxFrequency, count[s[r]]);

            while ((r - l + 1) - maxFrequency > k){
                count[s[l]]--;
                l++;
            }

            res = max(res, r-l+1);
        }


        return res;
    }
};
